### PR TITLE
[SC-5] Router and Market Breakout - Code Freeze Prep

### PR DESCRIPTION
**Pull request by mrtpain**
Breakout the router functionality from the market. Allows for updated routers to be deployed. 

Prepare for a code freeze during the audit.

### CODE REVIEW

File: src/Contracts/Deployer/OpdexMarketDeployer.cs [hunk 1]
```diff
@@ -30,57 +25,81 @@ public Address Owner
...
         const uint transactionFee = 3; // .3% for the staking market
+        
+        Assert(State.IsContract(stakingToken), "OPDEX: INVALID_STAKING_TOKEN");
 
-        var response = Create<OpdexStakingMarket>(0, new object[] {stakingToken, transactionFee});
+        var marketParams = new object[] {transactionFee, stakingToken};
+        var marketResponse = Create<OpdexStakingMarket>(0, marketParams);
         
-        Assert(response.Success, InvalidMarketError);
+        Assert(marketResponse.Success && marketResponse.NewContractAddress != Address.Zero, "OPDEX: INVALID_MARKET");
```
Comment (drmathias) [inline src/Contracts/Deployer/OpdexMarketDeployer.cs]: Surely if a `Create` call returns success then `NewContractAddress != Address.Zero` would _always_ be true.

File: src/Contracts/Deployer/OpdexMarketDeployer.cs [hunk 2]
```diff
@@ -30,57 +25,81 @@ public Address Owner
...
-        
-        var response = Create<OpdexStandardMarket>(0, new object[] {marketOwner, authPoolCreators, authProviders, authTraders, fee});
-        
-        Assert(response.Success, InvalidMarketError);
+        EnsureOwnerOnly();
+
+        var marketParams = new object[] {transactionFee, marketOwner, authPoolCreators, authProviders, authTraders, enableMarketFee};
+        var marketResponse = Create<OpdexStandardMarket>(0, marketParams);
+
+        Assert(marketResponse.Success && marketResponse.NewContractAddress != Address.Zero, "OPDEX: INVALID_MARKET");
```
Comment (drmathias) [inline src/Contracts/Deployer/OpdexMarketDeployer.cs:43]: See previous comment

File: src/Contracts/Deployer/OpdexMarketDeployer.cs [hunk 3]
```diff
@@ -30,57 +25,81 @@ public Address Owner
...
         });
+
+        return market;
+    }
+
+    private Address CreateOpdexRouter(Address market, uint transactionFee, bool authProviders, bool authTraders)
+    {
+        var routerResponse = Create<OpdexRouter>(0, new object[] {market, transactionFee, authProviders, authTraders});
+        
+        Assert(routerResponse.Success && routerResponse.NewContractAddress != Address.Zero, "OPDEX: INVALID_ROUTER");
```
Comment (drmathias) [inline src/Contracts/Deployer/OpdexMarketDeployer.cs:96]: See previous comment

File: src/Contracts/Routers/OpdexRouter.cs [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+    
+    /// <inheritdoc />
+    public bool AuthProviders
+    {
+        get => State.GetBool(nameof(AuthProviders));
+        private set => State.SetBool(nameof(AuthProviders), value);
+    }
+    
+    /// <inheritdoc />
+    public Address GetStoredPool(Address token)
```
Comment (drmathias) [inline src/Contracts/Routers/OpdexRouter.cs]: Nitpick: consider renaming to `GetPool` to match market contract

File: src/Contracts/Deployer/OpdexMarketDeployer.cs (hunk 1 above)
Comment (drmathias) [reply src/Contracts/Deployer/OpdexMarketDeployer.cs]: See this throughout the contracts

File: src/Contracts/Markets/OpdexStandardMarket.cs [hunk 5]
```diff
@@ -133,73 +134,34 @@ public override Address CreatePool(Address token)
...
-        return AddLiquidityExecute(token, amountSrcDesired, amountCrsMin, amountSrcMin, to, deadline);
-    }
-    
+
     /// <inheritdoc />
-    public override UInt256[] RemoveLiquidity(Address token, UInt256 liquidity, ulong amountCrsMin, UInt256 amountSrcMin, Address to, ulong deadline)
+    public void CollectMarketFees(Address token, UInt256 amount)
     {
-        EnsureAuthorizationFor(Message.Sender, Permissions.Provide);
+        if (!MarketFeeEnabled || amount == 0) return;
```
Comment (drmathias) [inline src/Contracts/Markets/OpdexStandardMarket.cs]:
```suggestion
        if (amount == 0 || !MarketFeeEnabled) return;
```
Really minor and insignificant, could save the tiniest amount of gas

File: src/Contracts/Deployer/OpdexMarketDeployer.cs (hunk 1 above)
Comment (mrtpain) [reply src/Contracts/Deployer/OpdexMarketDeployer.cs]: This is resolved in both repos now.

**Pull request merged**